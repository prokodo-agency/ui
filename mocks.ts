import fs from "fs"
import path from "path"

const animationsDir = path.join(__dirname, "src", "__mocks__", "animations")
const originalAnimationsDir = path.join(
  __dirname,
  "public",
  "json",
  "animations",
)

// Ensure the animations directory exists
if (!fs.existsSync(animationsDir)) {
  fs.mkdirSync(animationsDir, { recursive: true })
}

// Read the original animations directory and create mock files for each JSON file
fs.readdirSync(originalAnimationsDir).forEach(file => {
  const mockFilePath = path.join(animationsDir, file)
  if (!fs.existsSync(mockFilePath)) {
    fs.writeFileSync(mockFilePath, "{}")
  }
})

console.log("Manual mocks for animations created successfully.")
