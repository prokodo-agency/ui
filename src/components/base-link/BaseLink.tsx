"use client"
import { type FC, type CSSProperties } from "react"

import { isNull } from "@/helpers/validations"

import type { BaseLinkProps } from "./BaseLink.model"

export const BaseLink: FC<BaseLinkProps> = ({
  href,
  children,
  disabled,
  linkComponent: LinkComponent,
  style,
  target,
  rel,
  ...rest
}) => {
  const DISABLE_STYLES: CSSProperties = { pointerEvents: "none" }
  const isAbsolute = href?.startsWith("http")
  const isDownload = rest.download !== undefined
  const isAppUrl = /^\w+:/.test(href)

  const computedTarget = target ?? (isAbsolute ? "_blank" : undefined)
  const computedRel = rel ?? (isAbsolute ? "noopener noreferrer" : undefined)
  const tabIndex = !isNull(disabled) ? -1 : undefined
  const linkStyle = !isNull(disabled) ? { ...DISABLE_STYLES, ...style } : style

  // Use plain <a> for absolute or special URLs
  if (isAbsolute || isDownload || isAppUrl || LinkComponent === undefined) {
    return (
      <a
        {...rest}
        href={href}
        rel={computedRel}
        style={linkStyle}
        tabIndex={tabIndex}
        target={computedTarget}
      >
        {children}
      </a>
    )
  }

  // Internal link with custom component
  return (
    <LinkComponent {...rest} href={href} style={linkStyle} tabIndex={tabIndex}>
      {children}
    </LinkComponent>
  )
}

BaseLink.displayName = "BaseLink"
