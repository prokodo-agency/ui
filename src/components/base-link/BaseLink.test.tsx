import { expect } from "@jest/globals"

import { render } from "@/tests"
import { mockRouter } from "@/tests/mocks/router"

import { BaseLink } from "./BaseLink"

const locales = ["de", "en"]

describe("The common BaseLink component", () => {
  for (const locale of locales) {
    describe(`using locale '${locale}'`, () => {
      beforeEach(() => {
        mockRouter({
          back: jest.fn(),
          forward: jest.fn(),
          push: jest.fn(),
          replace: jest.fn(),
          refresh: jest.fn(),
          prefetch: jest.fn(),
        })
      })

      it("should render internal links", () => {
        const { container } = render(<BaseLink href="/">Home</BaseLink>)
        expect(container.firstElementChild).toMatchSnapshot()
      })

      it("should render external links", () => {
        const { container } = render(
          <BaseLink href="https://example.com">https://example.com</BaseLink>,
        )
        expect(container.firstElementChild).toMatchSnapshot()
      })

      it("should render download links", () => {
        const { container } = render(
          <BaseLink
            download="sample.pdf"
            href="http://www.africau.edu/images/default/sample.pdf"
          >
            sample.pdf
          </BaseLink>,
        )
        expect(container.firstElementChild).toMatchSnapshot()
      })

      it("should render disabled links", () => {
        const { container } = render(
          <BaseLink disabled href="https://example.com">
            https://example.com
          </BaseLink>,
        )
        expect(container.firstElementChild).toMatchSnapshot()
      })

      it("should render locale links", () => {
        const { container } = render(
          <BaseLink href="/" locale={locale}>
            Home
          </BaseLink>,
        )
        expect(container.firstElementChild).toMatchSnapshot()
      })

      it("should render mailto links", () => {
        const { container } = render(
          <BaseLink href="mailto:foo@example.com">foo@example.com</BaseLink>,
        )
        expect(container.firstElementChild).toMatchSnapshot()
      })

      it("should render anchor links", () => {
        const { container } = render(<BaseLink href="#bottom">Bottom</BaseLink>)
        expect(container.firstElementChild).toMatchSnapshot()
      })
    })
  }
})
