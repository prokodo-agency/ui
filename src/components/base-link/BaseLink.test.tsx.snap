// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The common BaseLink component using locale 'de' should render anchor links 1`] = `
<a
  href="#bottom"
>
  Bottom
</a>
`;

exports[`The common BaseLink component using locale 'de' should render disabled links 1`] = `
<a
  href="https://example.com"
  rel="noopener noreferrer"
  style="pointer-events: none;"
  tabindex="-1"
  target="_blank"
>
  https://example.com
</a>
`;

exports[`The common BaseLink component using locale 'de' should render download links 1`] = `
<a
  download="sample.pdf"
  href="http://www.africau.edu/images/default/sample.pdf"
  rel="noopener noreferrer"
  target="_blank"
>
  sample.pdf
</a>
`;

exports[`The common BaseLink component using locale 'de' should render external links 1`] = `
<a
  href="https://example.com"
  rel="noopener noreferrer"
  target="_blank"
>
  https://example.com
</a>
`;

exports[`The common BaseLink component using locale 'de' should render internal links 1`] = `
<a
  href="/"
>
  Home
</a>
`;

exports[`The common BaseLink component using locale 'de' should render locale links 1`] = `
<a
  href="/"
>
  Home
</a>
`;

exports[`The common BaseLink component using locale 'de' should render mailto links 1`] = `
<a
  href="mailto:foo@example.com"
>
  foo@example.com
</a>
`;

exports[`The common BaseLink component using locale 'en' should render anchor links 1`] = `
<a
  href="#bottom"
>
  Bottom
</a>
`;

exports[`The common BaseLink component using locale 'en' should render disabled links 1`] = `
<a
  href="https://example.com"
  rel="noopener noreferrer"
  style="pointer-events: none;"
  tabindex="-1"
  target="_blank"
>
  https://example.com
</a>
`;

exports[`The common BaseLink component using locale 'en' should render download links 1`] = `
<a
  download="sample.pdf"
  href="http://www.africau.edu/images/default/sample.pdf"
  rel="noopener noreferrer"
  target="_blank"
>
  sample.pdf
</a>
`;

exports[`The common BaseLink component using locale 'en' should render external links 1`] = `
<a
  href="https://example.com"
  rel="noopener noreferrer"
  target="_blank"
>
  https://example.com
</a>
`;

exports[`The common BaseLink component using locale 'en' should render internal links 1`] = `
<a
  href="/"
>
  Home
</a>
`;

exports[`The common BaseLink component using locale 'en' should render locale links 1`] = `
<a
  href="/"
>
  Home
</a>
`;

exports[`The common BaseLink component using locale 'en' should render mailto links 1`] = `
<a
  href="mailto:foo@example.com"
>
  foo@example.com
</a>
`;
